Sources/Overlays/_Testing_Foundation/Attachments/Attachable+Encodable.swift

```diff
@@ -0,0 +1,97 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+//
+
+#if canImport(Foundation)
+@_spi(Experimental) public import Testing
+private import Foundation
+
+/// A common implementation of ``withUnsafeBufferPointer(for:_:)`` that is
+/// used when a type conforms to `Encodable`, whether or not it also conforms
+/// to `NSSecureCoding`.
+///
+/// - Parameters:
+///   - attachment: The attachment that is requesting a buffer (that is, the
+///     attachment containing this instance.)
+///   - body: A function to call. A temporary buffer containing a data
+///     representation of this instance is passed to it.
+///
+/// - Returns: Whatever is returned by `body`.
+///
+/// - Throws: Whatever is thrown by `body`, or any error that prevented the
+///   creation of the buffer.
+func withUnsafeBufferPointer<E, R>(encoding attachableValue: borrowing E, for attachment: borrowing Attachment<E>, _ body: (UnsafeRawBufferPointer) throws -> R) throws -> R where E: Attachable & Encodable {
+  let format = try EncodingFormat(for: attachment)
+
+  let data: Data
+  switch format {
+  case let .propertyListFormat(propertyListFormat):
+    let plistEncoder = PropertyListEncoder()
+    plistEncoder.outputFormat = propertyListFormat
+    data = try plistEncoder.encode(attachableValue)
+  case .default:
+    // The default format is JSON.
+    fallthrough
+  case .json:
+    // We cannot use our own JSON encoding wrapper here because that would
+    // require it be exported with (at least) package visibility which would
+    // create a visible external dependency on Foundation in the main testing
+    // library target.
+    data = try JSONEncoder().encode(attachableValue)
```

Comment:
(Not for this PR, but for consideration as work progresses, in case you hadn't already thought of this…)

I think it'd be nice if there were some affordance which allowed users to _optionally_ customize the `JSONEncoder` and modify its encoding options.

Reply:
That'll be part of the `Metadata` PR.